src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -133,16 +43,14 @@ private unsafe ValueTask<int> ReadAsyncInternal(Memory<byte> destination, Cancel
                 ThrowHelper.ThrowNotSupportedException_UnreadableStream();
             }
 
-            // Rent the reusable ValueTaskSource, or create a new one to use if we couldn't get one (which
-            // should only happen on first use or if the FileStream is being used concurrently).
-            ValueTaskSource vts = Interlocked.Exchange(ref _reusableValueTaskSource, null) ?? new ValueTaskSource(this);
+            SafeFileHandle.ValueTaskSource vts = _fileHandle.GetValueTaskSource();
```

Comment:
We can't just replace a bunch of this code with a call to RandomAccess?  Same for other places.

Reply:
It is definitely possible, however, I was not able to come up with a clean solution.

There are two main differences:

* the strategy sets the current position to the previous value on error:

https://github.com/dotnet/runtime/blob/eaa3dffed58c513691df6ceeba9c8c1087137e4c/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs#L192

https://github.com/dotnet/runtime/blob/eaa3dffed58c513691df6ceeba9c8c1087137e4c/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs#L278-L284

* since `RandomAccess` is supposed to be thread-safe and stateless, `Write` methods don't return `VT` but `VT<int>` instead. I could map `VT<int>` to `VT` but this would require casting and increase the memory allocations

I had few drafts locally, but I was happy about none of them and I currently gave up on that.

Reply:
> I could map VT<int> to VT but this would require casting and increase the memory allocations

This code is in corelib.  You can make the object used to back the ValueTask implement both `IValueTaskSource` and `IValueTaskSource<int>`, and add an internal helper to ValueTask that does the cast without allocation.  This is something we've considered exposing publicly, and only haven't because there's no guarantee that something implementing `IValueTaskSource<T>` also implements `IValueTaskSource`, which would make the implementation allocate a wrapper object in that case.  We still might do so, though.  Regardless, it'd be fine to add internally for now if you need it.